test/parse.jl

```diff
@@ -326,3 +305,19 @@ end
     tmp = @. [Int, Float64, String, Bool] <: Union{Int, String}
     @test tmp == Bool[1, 0,1, 0]
 end
+
+@testset "tryparse int on iterators" begin
+    s = Iterators.filter(!=('_'), "1_2_3_4")
+    @test Base.Parse.tryparse_internal_int(Int, s, 10, true) == 1234
```

Comment:
This is a bit sneaky. My original intention with this PR was to be able to run `parse` on an arbitrary iterator (to support parsing `"1_2_3_4` as an integer which the TOML specification allows using `Iterators.filter`) and it happens that with the refactoring the part that does the int parsing is now general enough to allow this).